sdk/iOS/test/MSSyncTableTests.m

```diff
@@ -1082,9 +1083,348 @@ -(void) testPushNetworkTimeoutWithErroredOps
+    __block NSOperation *pullOperation;
+    
+    // Define filters
+    ((MSTestFilter *)filter.testFilters[0]).onInspectRequest = ^(NSURLRequest *request) {
+        
+        // Cancellation point 0
+        if (requestedCancellationPoint == 0)
+            [pullOperation cancel];
+        
+        return request;
+    };
+    
+    ((MSTestFilter *)filter.testFilters[0]).onInspectResponseData = ^(NSURLRequest *request, NSData *data) {
+        
+        // Cancellation point 1
+        if (requestedCancellationPoint == 1)
+            [pullOperation cancel];
+        
+        return data;
+    };
+    
+    ((MSTestFilter *)filter.testFilters[1]).onInspectRequest = ^(NSURLRequest *request) {
+        
+        // Cancellation point 2
+        if (requestedCancellationPoint == 2)
+            [pullOperation cancel];
+        
+        return request;
+    };
+    
+    ((MSTestFilter *)filter.testFilters[1]).onInspectResponseData = ^(NSURLRequest *request, NSData *data) {
+        
+        // Cancellation point 3
+        if (requestedCancellationPoint == 3)
+            [pullOperation cancel];
+        
+        return data;
+    };
+    
+    // Insert an item
+    XCTestExpectation *insertExpectation = [self expectationWithDescription:@"insert expectation"];
+    [todoTable insert:@{ @"id" : @"id1", @"text" : @"client" } completion:^(NSDictionary *item, NSError *error) {
+        [insertExpectation fulfill];
+    }];
+    
+    [self waitForExpectationsWithTimeout:10 handler:nil];
+    
+    // Perform a pull
+    pullOperation = [filteredClient.syncContext pullWithQuery:query queryId:nil completion:^(NSError *error) {
+
+        // Cancellation point 4
+        if (requestedCancellationPoint == 4)
+            [pullOperation cancel];
+    }];
+
+    // Cancellation point 5
+    if (requestedCancellationPoint == 5)
+        [pullOperation cancel];
+
+    [self verifyFinished:pullOperation];
+    
+    int synchronizedItemCount = [self synchronizedItemCount];
+    
+    XCTAssertEqual([self performVanillaPull], 3, "Expected all server items to have been pulled to the client by now");
+    
+    return synchronizedItemCount;
+}
+
+// Performs a pull operation and returns the count of all the server items pulled so far
+-(int) performVanillaPull
+{
+    // Initialization
+    MSTestFilter *filter = [MSTestFilter new];
+    MSClient *filteredClient = [client clientWithFilter:filter];
+    MSSyncTable *todoTable = [filteredClient syncTableWithName:TodoTableNoVersion];
+    MSQuery *query = [[MSQuery alloc] initWithSyncTable:todoTable];
+
+    __block bool isFirstPullRequest = true;
+    filter.ignoreNextFilter = YES;
+    filter.onInspectResponseData = ^(NSURLRequest *request, NSData *data) {
+
+        NSDictionary *item = [filteredClient.serializer itemFromData:request.HTTPBody withOriginalItem:nil ensureDictionary:YES orError:nil];
+        
+        NSString *responseString;
+        if (item != nil) // This means it is a push request
```

Comment:
There is a difference in what the first filter does and what the filter in performVanillaPush (method renamed now) does.
The first filter expects a push request for the inserted item followed by a pull request and is statically configured to respond accordingly.
The second filter may get variable number of push requests depending on when the earlier pull was cancelled. This filter looks at the requests and responds accordingly dynamically. This is the reason for overriding the first filter.